### ISSUE TITLE
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder"

### ISSUE THREAD

**Issue by Flo354** [playframework/playframework#7422]
### Play Version (2.5.x / etc)
2.6.0-RC2

### API (Scala / Java / Neither / Both)
JAVA

### Operating System (Ubuntu 15.10 / MacOS 10.10 / Windows 10)
Linux Florian-Ubuntu 4.4.0-78-generic #99-Ubuntu SMP Thu Apr 27 15:29:09 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux

### JDK (Oracle 1.8.0_72, OpenJDK 1.8.x, Azul Zing)
Oracle, OpenJDK

java version "1.8.0_131"
Java(TM) SE Runtime Environment (build 1.8.0_131-b11)
Java HotSpot(TM) 64-Bit Server VM (build 25.131-b11, mixed mode)

### Expected Behavior

1. Starting play app
2. Loading logger
3. Logger loaded and output according to configuration

### Actual Behaviour

1. Starting play app
2. Loading logger
3. SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder"
4. Output every single log

Getting Scala 2.10.6 (for sbt)...
:: retrieving :: org.scala-sbt#boot-scala
	confs: [default]
	5 artifacts copied, 0 already retrieved (24494kB/247ms)
[info] Loading project definition from /home/florian/IdeaProjects/XXX/project
[info] Set current project to XXX (in build file:/home/florian/IdeaProjects/XXX/)
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.

### EDIT

I said something wrong. In the "Actual Behaviour" part, I said "4. Output every single log".
It's not true, I just forgot to disable debug loggin for the new play-ws.
But there is still the failure while loading the class StaticLoggerBinder.

**Comment by jjteo74** [comment]
Hit the same problem when using IntelliJ. Added `logback-classic` to `plugins.sbt` as a work around

```
// The Play plugin
addSbtPlugin("com.typesafe.play" % "sbt-plugin" % "2.6.0-RC2")
// workaround for missing static SLF4J binder for logback
libraryDependencies += "ch.qos.logback" % "logback-classic" % "1.2.3"
```

**Comment by fmeriaux** [comment]
Same for me. Adding dependency is the final solution? That seems surprising to me

**Comment by jjteo74** [comment]
Apparently that message is from sbt plugin??? `logback.xml` is still being picked up without the workaround: -

```
--- (Running the application, auto-reloading is enabled) ---

[info] p.c.s.AkkaHttpServer - Listening for HTTP on /0:0:0:0:0:0:0:0:9000

Server started, use Cmd+P to stop

SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
18:28:38.971 [scala-execution-context-global-72] [warn] application - application.conf @ file:/xxx/target/scala-2.12/classes/application.conf: 4: play.crypto.secret is deprecated, use play.http.secret.key instead
18:28:39.331 [play-dev-mode-akka.actor.default-dispatcher-3] [warn] application - application.conf @ file:/xxx/target/scala-2.12/classes/application.conf: 4: play.crypto.secret is deprecated, use play.http.secret.key instead
18:28:39.425 [play-dev-mode-akka.actor.default-dispatcher-3] [warn] p.a.m.DefaultUrlEncodedCookieDataCodec - Cookie failed message authentication check
```

**Comment by Flo354** [comment]
@jjteo74 yes you are right, I edited my post yesterday to say that even if there is an error shown, the file is loaded!

I agree with @fmeriaux, don't we have another solution to avoid adding a dependency?

**Comment by wsargent** [comment]
This error messages happens when you have the SLF4J API loaded, but do not have an SLF4J implementation (i.e. logback, log4j2, etc) loaded.  The logs you're seeing are from Akka, which keeps going even if there is not an SLF4J implementation loaded.

Unfortunately, there's not enough information here to show how the project got to that state. Can you attach your build.sbt, projects/plugins.sbt and application.conf?

**Comment by Flo354** [comment]
Yes!

application.conf
```scala
application.environment = development

play.application.loader = "configuration.CustomApplicationLoader"

akka {
  log-config-on-start = false
}

play.modules {
  enabled += modules.formatters.FormattersModule
  enabled += modules.jwt.JWTModule
  enabled += modules.redisson.RedissonModule
  enabled += modules.testregistration.RemoveTestRegistrationMobule

  disabled += "play.data.format.FormattersModule"
  #disabled += "play.api.db.DBModule"
  #enabled += configuration.modules.DBModule
}

play.i18n {
  langs = [ "en" ]
}

play.http {
  #errorHandler = configuration.CustomErrorHandler
  filters = configuration.filters.Filters
  secret.key = "my_key"
}

play.filters {
  cors {
    allowedHttpMethods = ["GET", "PUT"]
  }

  hosts {
    allowed = [".arcanaccess.com", "localhost:9000", "localhost"]
  }
}

play.dbprototype.driver = com.mysql.cj.jdbc.Driver

play.db.default = "default"
play.evolutions {
  enabled = true
  useLocks = true
  autoApply = true
}

ebean.default = "mysql.models.*"
```

build.properties
```scala
#Activator-generated Properties
#Sat Jun 25 14:18:31 CEST 2016
template.uuid=cb6aec74-d62b-4c96-9c67-a5d023dbdc00
sbt.version=0.13.15
```

build.sbt
```scala
scalaVersion := "2.12.2"

resolvers += "jitpack" at "https://jitpack.io"
resolvers += Resolver.sonatypeRepo("snapshots")

lazy val common = (project in file("common")).
  enablePlugins(PlayJava, PlayEbean)

lazy val aa_api = (project in file("."))
  .enablePlugins(PlayJava, PlayEbean, SbtWeb)
  .dependsOn(common)
  .aggregate(common)

libraryDependencies ++= Seq(
  ehcache,
  filters,
  guice,
  ws,

  "org.jetbrains" % "annotations" % "15.0",

  "com.github.fdimuccio" % "play2-sockjs_2.12" % "0.6.0-SNAPSHOT",
  "org.redisson" % "redisson" % "3.4.2",

  "mysql" % "mysql-connector-java" % "6.0.6",
  "org.hibernate.javax.persistence" % "hibernate-jpa-2.1-api" % "1.0.0.Final",

  "org.bouncycastle" % "bcprov-jdk15on" % "1.56",
  "com.auth0" % "java-jwt" % "3.1.0",

  "com.amazonaws" % "aws-java-sdk-ses" % "1.11.132",
  "com.amazonaws" % "aws-java-sdk-ec2" % "1.11.132",

  "com.googlecode.libphonenumber" % "libphonenumber" % "8.5.0",
  "commons-validator" % "commons-validator" % "1.6"
)
....
```

plugins.sbt
```scala
// The Play plugin
addSbtPlugin("com.typesafe.play" % "sbt-plugin" % "2.6.0-RC2")
addSbtPlugin("com.typesafe.sbt" % "sbt-play-ebean" % "4.0.0-M3")

addSbtPlugin("com.slidingautonomy.sbt" % "sbt-filter" % "1.0.1")
// addSbtPlugin("com.typesafe.sbt" % "sbt-uglify" % "1.0.3")
// replaced because of exception, more info here: https://github.com/sbt/sbt-uglify/issues/25
addSbtPlugin("org.danielnixon" % "sbt-uglify" % "1.0.7")
addSbtPlugin("net.ground5hark.sbt" % "sbt-css-compress" % "0.1.4")
addSbtPlugin("net.ground5hark.sbt" % "sbt-concat" % "0.1.9")
```

subproject common: build.sbt
```scala
scalaVersion := "2.12.2"

libraryDependencies ++= Seq(
  "com.fasterxml.jackson.core" % "jackson-core" % "2.8.8",
  "com.fasterxml.jackson.core" % "jackson-databind" % "2.8.8.1",
  "com.fasterxml.jackson.core" % "jackson-annotations" % "2.8.8",

  "org.hibernate.javax.persistence" % "hibernate-jpa-2.1-api" % "1.0.0.Final",

  "org.bouncycastle" % "bcprov-jdk15on" % "1.56",

  "com.amazonaws" % "aws-java-sdk-ses" % "1.11.132",
  "com.amazonaws" % "aws-java-sdk-ec2" % "1.11.132"
)
....
```

**Comment by PromanSEW** [comment]
I have the same problem

**Comment by rmcloughlin** [comment]
One easy way to reproduce this problem is as follows:

```
$ sbt new playframework/play-java-seed.g8
$ cd play-java-seed/
$ sbt test
```

**Comment by wsargent** [comment]
Okay, verified.  Poking at this for a bit...

**Comment by wsargent** [comment]
The flow is: 

```
DevServerStart
LogbackLoggerConfigurator
play.api.Logger.setApplication
object Logger instantiation 
call to LoggerFactory.getLogger("application")
```

Because Logger has a call in the constructor:

```
object Logger extends Logger(LoggerFactory.getLogger("application")) {
```

https://github.com/playframework/playframework/blob/master/framework/src/play/src/main/scala/play/api/Logger.scala#L276

So because LoggerFactory is called here before the configuration, there is no binding in place for the LoggerFactory and it errors at the `bind`.

This looks like it's coming from https://github.com/playframework/playframework/pull/6832

**Comment by wsargent** [comment]
This is also happening from SbtWeb, because `org.webjars.WebjarExtractor` is also calling out to LoggerFactory.

This shouldn't be happening, because PlayLogback auto plugin should have the logback classes in scope.  Will look to see if the classloaders have changed recently.

**Comment by wsargent** [comment]
It's not WebjarExtractor, because that was 3 years ago.  Creating a sample project and working through the milestones, I don't see this in 2.6.0-M5 but I do see it in 2.6.0-RC1.

**Comment by jamesward** [comment]
My initial guess was that the Play sbt plugin no longer transitively depends on `logback-classic` and that is only in the root project dependency chain.  But I haven't dug deeper.

**Comment by Flo354** [comment]
I don't think either that the problem is from Webjar since I don't use it in my project.
I switched directly from 2.5.x to 2.6.0-RC2. The problem is still here in 2.6.0

**Comment by wsargent** [comment]
Exists after https://github.com/playframework/playframework/commit/dbed4489ea0b2dee946aaf7585b83f4a70b64c87

**Comment by wsargent** [comment]
Does not exist in eb149b4bbda594437c1e294910edc0328a9155fa and applying https://github.com/playframework/playframework/pull/7349 by  upgrading to sbt-js-engine to 1.2.0 by hand shows the error.  So it's somewhere in sbt-js-engine...

**Comment by wsargent** [comment]
It looks like sbt-js-engine got an upgrade to sbt-web 2.4.0 between 1.1.4 and 1.2.0 though: https://github.com/sbt/sbt-js-engine/commit/eafaa485b6aca8138802e8614b8f7dcfb759af7e#diff-fdc3abdfd754eeb24090dbd90aeec2ceR13

**Comment by wsargent** [comment]
Looking at https://github.com/sbt/sbt-js-engine/commits/1.2.0 doesn't show anything that stands out as a big flag.  And in point of fact we don't know that the bug was in that commit -- the bug could have been earlier, and it could have simply been the thing that revealed it.

**Comment by wsargent** [comment]
SBT-Plugin as of eb149b4bbda594437c1e294910edc0328a9155fa

```
> dependencyList
[warn] Credentials file /home/wsargent/.bintray/.credentials does not exist
[info] Updating {file:/home/wsargent/work/playframework/framework/}SBT-Plugin...
[info] Resolving org.scala-sbt.ivy#ivy;2.3.0-sbt-48dd0744422128446aee9ac31aa356e[info] Resolving org.scala-sbt#sbt-launch;0.13.15 ...
[info] Done updating.
[info] org.slf4j:slf4j-api:1.7.12
[info] org.slf4j:slf4j-simple:1.7.12
```

SBT-Plugin with the upgrade to sbt-js-engine:

```
> dependencyList
[warn] Credentials file /home/wsargent/.bintray/.credentials does not exist
[info] args4j:args4j:2.0.26
[info] com.fasterxml.jackson.core:jackson-annotations:2.3.0
[info] com.fasterxml.jackson.core:jackson-core:2.7.3
[info] com.fasterxml.jackson.core:jackson-databind:2.3.3
[info] com.github.pathikrit:better-files_2.10:2.14.0
[info] com.google.code.findbugs:jsr305:1.3.9
[info] com.google.code.gson:gson:2.2.4
[info] com.google.guava:guava:19.0
[info] com.google.protobuf:protobuf-java:2.5.0
[info] com.googlecode.json-simple:json-simple:1.1.1
[info] com.lightbend.play:jnotify:0.94-play-2
[info] com.lightbend.play:play-file-watch_2.10:1.0.0
[info] com.typesafe:config:1.3.1
[info] com.typesafe:jse_2.10:1.2.2
[info] com.typesafe:npm_2.10:1.2.0
[info] com.typesafe.akka:akka-actor_2.10:2.3.16
[info] com.typesafe.akka:akka-cluster_2.10:2.3.16
[info] com.typesafe.akka:akka-contrib_2.10:2.3.16
[info] com.typesafe.akka:akka-persistence-experimental_2.10:2.3.16
[info] com.typesafe.akka:akka-remote_2.10:2.3.16
[info] com.typesafe.play:build-link:2.6.0-SNAPSHOT
[info] com.typesafe.play:play-exceptions:2.6.0-SNAPSHOT
[info] com.typesafe.play:run-support_2.10:2.6.0-SNAPSHOT
[info] com.typesafe.play:sbt-plugin:2.6.0-SNAPSHOT
[info] com.typesafe.play:sbt-routes-compiler_2.10:2.6.0-SNAPSHOT
[info] com.typesafe.play:twirl-api_2.10:1.3.0
[info] com.typesafe.play:twirl-compiler_2.10:1.3.0
[info] com.typesafe.play:twirl-parser_2.10:1.3.0
[info] com.typesafe.sbt:sbt-js-engine:1.2.0
[info] com.typesafe.sbt:sbt-native-packager:1.1.6
[info] com.typesafe.sbt:sbt-twirl:1.3.0
[info] com.typesafe.sbt:sbt-web:1.4.0
[info] commons-io:commons-io:2.5
[info] io.apigee:rhino:1.7R5pre4
[info] io.apigee.trireme:trireme-core:0.8.9
[info] io.apigee.trireme:trireme-kernel:0.8.9
[info] io.apigee.trireme:trireme-node10src:0.8.9
[info] io.netty:netty:3.8.0.Final
[info] io.spray:spray-json_2.10:1.3.3
[info] net.contentobjects.jnotify:jnotify:0.94-play-1
[info] org.apache.ant:ant:1.9.6
[info] org.apache.ant:ant-launcher:1.9.6
[info] org.apache.commons:commons-compress:1.9
[info] org.apache.commons:commons-lang3:3.4
[info] org.fusesource.hawtjni:hawtjni-runtime:1.8
[info] org.fusesource.leveldbjni:leveldbjni:1.7
[info] org.fusesource.leveldbjni:leveldbjni-all:1.7
[info] org.fusesource.leveldbjni:leveldbjni-linux32:1.5
[info] org.fusesource.leveldbjni:leveldbjni-linux64:1.5
[info] org.fusesource.leveldbjni:leveldbjni-osx:1.5
[info] org.fusesource.leveldbjni:leveldbjni-win32:1.5
[info] org.fusesource.leveldbjni:leveldbjni-win64:1.5
[info] org.iq80.leveldb:leveldb:0.5
[info] org.iq80.leveldb:leveldb-api:0.5
[info] org.mozilla:rhino:1.7.7.1
[info] org.scala-js:closure-compiler-externs-java-6:v20160517
[info] org.scala-js:closure-compiler-java-6:v20160517
[info] org.scala-js:sbt-scalajs:0.6.13
[info] org.scala-js:scalajs-ir_2.10:0.6.13
[info] org.scala-js:scalajs-js-envs_2.10:0.6.13
[info] org.scala-js:scalajs-sbt-test-adapter_2.10:0.6.13
[info] org.scala-js:scalajs-tools_2.10:0.6.13
[info] org.scala-lang:scala-compiler:2.10.6
[info] org.scala-lang:scala-reflect:2.10.6
[info] org.scala-sbt:test-interface:1.0
[info] org.slf4j:slf4j-api:1.7.22
[info] org.tukaani:xz:1.0
[info] org.uncommons.maths:uncommons-maths:1.2.2a
[info] org.webjars:envjs:1.2
[info] org.webjars:npm:4.2.0
[info] org.webjars:webjars-locator:0.26
[info] org.webjars:webjars-locator-core:0.32
[success] Total time: 1 s, completed Jun 28, 2017 2:21:34 PM
```

So no slf4j-simple in the mix.

**Comment by wsargent** [comment]
Turns out this is the other end of the sbt-js-engine bug from 2015: https://github.com/typesafehub/js-engine/issues/44

**Comment by richdougherty** [comment]
Hi @Flo354, this issue is now addressed in Play 2.6.1 if you want to try it out.

**Comment by Flo354** [comment]
Problem solved! Thanks,
